core/src/com/unciv/ui/utils/extensions/Scene2dExtensions.kt

```diff
@@ -62,6 +69,183 @@ fun Color.darken(t: Float): Color = Color(this).lerp(Color.BLACK, t)
+
+    val keyShortcuts: ActorKeyShortcutDispatcher
+
+    private constructor(actor: Actor) {
+        keyShortcuts = ActorKeyShortcutDispatcher(actor)
+    }
+
+    fun activate() {
+        if (this::activationActions.isInitialized) {
+            for (action in activationActions)
+                action()
+        }
+    }
+
+    fun addActivationAction(action: () -> Unit) {
+        if (!this::activationActions.isInitialized) activationActions = mutableListOf()
+        activationActions.add(action)
+
+        if (clickActivationListener == null) {
+            clickActivationListener = object: ClickListener() {
+                override fun clicked(event: InputEvent?, x: Float, y: Float) {
+                    actor.activate()
+                }
+            }
+            actor.addListener(clickActivationListener)
+        }
+    }
+
+    fun removeActivationAction(action: () -> Unit) {
+        if (!this::activationActions.isInitialized) return
+        activationActions.remove(action)
+        if (activationActions.none() && clickActivationListener != null) {
+            actor.removeListener(clickActivationListener)
+            clickActivationListener = null
+        }
+    }
+}
+
+
+fun Actor.addActivationAction(action: (() -> Unit)?) {
+    if (action != null)
+        ActorAttachments.get(this).addActivationAction(action)
+}
+
+fun Actor.removeActivationAction(action: (() -> Unit)?) {
+    if (action != null)
+        ActorAttachments.getOrNull(this)?.removeActivationAction(action)
+}
+
+fun Actor.isActive(): Boolean = isVisible() && !(this is Disableable && (this as Disableable).isDisabled())
+
+fun Actor.activate() {
+    if (isActive())
+        ActorAttachments.getOrNull(this)?.activate()
+}
+
+val Actor.keyShortcutsOrNull
+    get() = ActorAttachments.getOrNull(this)?.keyShortcuts
+val Actor.keyShortcuts
+    get() = ActorAttachments.get(this).keyShortcuts
+
+fun Actor.onActivation(sound: UncivSound = UncivSound.Click, action: () -> Unit): Actor {
+    addActivationAction({
+        launchCrashHandling("Sound") { SoundPlayer.play(sound) }
+        action()
+    })
+    return this
+}
+
+fun Actor.onActivation(action: () -> Unit): Actor = onActivation(UncivSound.Click, action)
+
+
+enum class DispatcherVetoResult { Accept, Skip, SkipWithChildren }
+typealias DispatcherVetoer = (associatedActor: Actor?, keyDispatcher: KeyShortcutDispatcher?) -> DispatcherVetoResult
+
+/**
+ * Install shorcut dispatcher for this stage. It activates all actions associated with the
+ * pressed key in [additionalShortcuts] (if specified) and all actors in the stage. It is
+ * possible to temporarily disable or veto some shortcut dispatchers by passing an appropriate
+ * [dispatcherVetoerCreator] function. This function may return a [DispatcherVetoer], which
+ * will then be used to evaluate all shortcut sources in the stage. This two-step vetoing
+ * mechanism allows the callback ([dispatcherVetoerCreator]) perform expensive preparations
```

Comment:
I created this taking into account that some screens have really many actors, e.g. the world screen has hundreds if not thousands. I haven't taken into account that they form a tree, i.e. there only ~15 toplevel actors. However, I propose to keep this optimization, so that shorcut resolution code doesn't suddenly become slow on screens with thousands of toplevel actors: it would be O(N²) for number of actors, because property `activePopup`, while a oneliner, is not quite O(1). It is "making the code a little harder to understand" just in one place, after all.